src/tree-sitter-language-mode.js

```diff
@@ -0,0 +1,506 @@
+const {Document} = require('tree-sitter')
+const {Point, Range, Emitter} = require('atom')
+const ScopeDescriptor = require('./scope-descriptor')
+const TokenizedLine = require('./tokenized-line')
+
+let nextId = 0
+
+module.exports =
+class TreeSitterLanguageMode {
+  constructor ({buffer, grammar, config}) {
+    this.id = nextId++
+    this.buffer = buffer
+    this.grammar = grammar
+    this.config = config
+    this.document = new Document()
+    this.document.setInput(new TreeSitterTextBufferInput(buffer))
+    this.document.setLanguage(grammar.languageModule)
+    this.document.parse()
+    this.rootScopeDescriptor = new ScopeDescriptor({scopes: [this.grammar.id]})
```

Comment:
This will break certain atom packages that depend on this being `this.grammar.scopeName` like the current [text-mate-language-mode](https://github.com/atom/atom/blob/8c92ed96525e29397b706510556b2caf54f572e4/src/text-mate-language-mode.js#L34).

Examples of packages that will break:
* [atom-typescript](https://github.com/TypeStrong/atom-typescript/blob/34072c896ebf8b54620b447139c2df0492f1faf0/lib/main/atom/utils/atom.ts#L24-L27)

Reply:
Yeah, we're intentionally using different-looking scope names because the whole idea of the dot-separated names like `source.ts` originates from TextMate and isn't our ideal API. In the long-term, we'd like for scoped settings to use simple language names like `typescript` instead.

We have added the concept of 'scope synonyms' to the Atom config system, so that in the meantime, you'll be able to apply scoped settings to `source.js` in your config system and it will apply to the TreeSitter grammar whose id is `javascript`.

I agree that this is going to break packages. Before we enable the `core.useTreeSitterParsers` setting by default, we'll need to figure out more of a solution for this. Most likely, it will involve a combination of API additions and also opening large numbers of pull requests on community packages to help keep everyone's code working.